### ISSUE TITLE
Compatibility Issue: Haptic module code error

### ISSUE THREAD

**Issue by q1221sr** [sofa-framework/sofa#4336]
i want to compile Haptic module, but The compiler tells me that these functions are inaccessible
(vs 2022  windows 10 sdk sofa 22.12)

visualHaptionDOF = new sofa::component::container::MechanicalObject<sofa::defaulttype::Rigid3dTypes>();
.
.
visualAxesDOF = new sofa::component::container::MechanicalObject<sofa::defaulttype::Rigid3dTypes>();
.
.
visualNode[i].visu = new sofa::component::visualmodel::OglModel();
.
.

and i find MechanicalObject and OglModel classes now have PROTECTED prefixes for constructors and destructors.
How do I change the code in the haptic module now?

**Comment by alxbilger** [comment]
Hi,

What is exactly the "Haptic module"? It's weird that is tries to use directly the constructors. You are right that the constructors are protected, so they cannot be used. So I am curious to know if this code ever worked....

To fix the code, you should have a look on the [tutorials](https://github.com/sofa-framework/sofa/tree/master/applications/tutorials), for example this line: https://github.com/sofa-framework/sofa/blob/master/applications/tutorials/oneTetrahedron/oneTetrahedron.cpp#L94
It uses the friend class `New` to instantiate the objects. There are also many examples in the unit tests.

**Comment by q1221sr** [comment]
> Hi,
> 
> What is exactly the "Haptic module"? It's weird that is tries to use directly the constructors. You are right that the constructors are protected, so they cannot be used. So I am curious to know if this code ever worked....
> 
> To fix the code, you should have a look on the [tutorials](https://github.com/sofa-framework/sofa/tree/master/applications/tutorials), for example this line: https://github.com/sofa-framework/sofa/blob/master/applications/tutorials/oneTetrahedron/oneTetrahedron.cpp#L94 It uses the friend class `New` to instantiate the objects. There are also many examples in the unit tests.

The code in plugins/Haption/HaptionDriver.cpp which support virtuose device
and i try to use the friend class ::New as

`visualHaptionDOF = sofa::component::container::MechanicalObject::New<sofa::defaulttype::Rigid3dTypes>();`

i'm sorry, still error

**Comment by alxbilger** [comment]
The right syntax is the following:

```C++
const auto visualHaptionDOFSPtr = sofa::core::objectmodel::New<sofa::component::statecontainer::MechanicalObject<sofa::defaulttype::Rigid3dTypes>>();
visualHaptionDOF = visualHaptionDOFSPtr.get();
nodeHaptionVisual->addObject(visualHaptionDOFSPtr);
visualHaptionDOF->name.setValue("rigidDOF");
```

Note that this plugin has not been maintained for a long time. You should expect some changes in order to make it work. Here are some changes I made today to fix the configuration: https://github.com/sofa-framework/sofa/pull/4338. But it does not fix the compilation.

If you succeed to make it work, a pull request would be very welcome ;)

**Comment by q1221sr** [comment]
> The right syntax is the following:
> 
> ```c++
> const auto visualHaptionDOFSPtr = sofa::core::objectmodel::New<sofa::component::statecontainer::MechanicalObject<sofa::defaulttype::Rigid3dTypes>>();
> visualHaptionDOF = visualHaptionDOFSPtr.get();
> nodeHaptionVisual->addObject(visualHaptionDOFSPtr);
> visualHaptionDOF->name.setValue("rigidDOF");
> ```
> 
> Note that this plugin has not been maintained for a long time. You should expect some changes in order to make it work. Here are some changes I made today to fix the configuration: #4338. But it does not fix the compilation.
> 
> If you succeed to make it work, a pull request would be very welcome ;)

We are in the process of adapting the sofa22.12 plugin for homebrew devices, with reference to the Haptic plugin. No problem of course, I'll commit our code as soon as it's aligned and tested.
